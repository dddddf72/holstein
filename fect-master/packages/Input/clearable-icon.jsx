import { defineComponent, computed } from 'vue'
import './clearable-icon.less'

const ClearableIcon = defineComponent({
  props: {
    visible: Boolean,
    disabled: Boolean,
  },
  emits: ['click'],
  setup(props, { emit }) {
    const clearableStyle = computed(() => {
      let str = ''
      props.visible && (str += ' visible')
      props.disabled && (str += ' disabled')
      return str
    })

    const handleClick = (e) => {
      e.preventDefault()
      e.stopPropagation()
      e.stopImmediatePropagation()
      emit('click', e)
    }

    return () => (
      <div
        className={`input__clear-icon ${clearableStyle.value}`}
        onClick={handleClick}
      >
        <svg
          viewBox="0 0 24 24"
          stroke="currentColor"
          strokeWidth="1.5"
          strokeLinecap="round"
          strokeLinejoin="round"
          fill="none"
          shapeRendering="geometricPrecision"
        >
          <path d="M18 6L6 18" />
          <path d="M6 6l12 12" />
        </svg>
      </div>
    )
  },
})

export default ClearableIcon
