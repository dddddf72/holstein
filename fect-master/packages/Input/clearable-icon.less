.fect-input {
  .input__clear-icon {
    padding: 0 var(--fay-gap-half);
    margin: 0;
    display: inline-flex;
    align-items: center;
    height: 100%;
    box-sizing: border-box;
    transition: color 150ms ease 0s;
    cursor: pointer;
    visibility: hidden;
    color: var(--accents-3);
    opacity: 0;
    &.visible {
      visibility: visible;
      opacity: 1;
    }
    &.disabled {
      cursor: not-allowed;
      pointer-events: none;
    }
    &:hover {
      color: var(--primary-foreground);
    }

    svg {
      color: currentColor;
      width: calc(var(--heightRatio) * 10.66px);
      height: calc(var(--heightRatio) * 10.66px);
    }
  }
}
