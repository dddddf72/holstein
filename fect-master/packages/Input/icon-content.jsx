import { defineComponent, computed } from 'vue'
import './icon-content.less'

const IconContent = defineComponent({
  props: {
    clickable: Boolean,
  },
  emits: ['click'],
  setup(props, { slots, emit }) {
    const baseStyle = computed(() => {
      const style = {
        cursor: !props.clickable ? 'pointer' : 'default',
        PointerEvetns: !props.clickable ? 'auto' : 'none',
      }
      return style
    })

    return () => (
      <div
        className="input-icon"
        style={baseStyle.value}
        onClick={(e) => emit('click', e)}
      >
        {slots.default?.()}
      </div>
    )
  },
})

export default IconContent
