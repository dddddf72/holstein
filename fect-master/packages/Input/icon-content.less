.fect-input {
  .input-icon {
    box-sizing: content-box;
    display: flex;
    align-items: center;
    vertical-align: middle;
    margin: 0;
    width: calc(var(--heightRatio) * 10.66px);
    padding: calc(var(--heightRatio) * var(--fay-gap) * 0.3);
    line-height: 1;
    position: relative;
  }
}
