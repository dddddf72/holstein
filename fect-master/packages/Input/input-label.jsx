import { defineComponent } from 'vue'
import './input-label.less'
const InputLabel = defineComponent({
  props: {
    fontSize: String,
    isRight: Boolean,
  },
  setup(props, { slots }) {
    return () => (
      <div
        class={`input-label ${props.isRight ? 'suffix' : ''}`}
        style={{ fontSize: props.fontSize }}
      >
        {slots.default?.()}
      </div>
    )
  },
})

export default InputLabel
