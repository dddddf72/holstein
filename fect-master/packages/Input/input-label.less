.fect-input {
  .input_container {
    .input-label {
      display: inline-flex;
      width: initial;
      height: 100%;
      align-items: center;
      pointer-events: none;
      margin: 0;
      padding: 0 var(--fay-gap-half);
      color: var(--accents-4);
      background-color: var(--accents-1);
      border-top-left-radius: var(--fay-radius);
      border-bottom-left-radius: var(--fay-radius);
      border-top: 1px solid var(--fay-border);
      border-left: 1px solid var(--fay-border);
      border-bottom: 1px solid var(--fay-border);
      line-height: 1;
      &.suffix {
        border-top-left-radius: 0;
        border-bottom-left-radius: 0;
        border-top-right-radius: var(--fay-radius);
        border-bottom-right-radius: var(--fay-radius);
        border-left: 0;
        border-right: 1px solid var(--fay-border);
      }
    }
  }
}
