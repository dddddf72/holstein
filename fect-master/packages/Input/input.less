.fect-input {
  width: inherit;
  display: inline-block;
  box-sizing: border-box;
  -webkit-box-align: center;
  .input_container {
    display: inline-flex;
    align-items: center;
    width: inherit;
    height: calc(var(--heightRatio) * var(--fay-gap));
  }
  .input_wrapper {
    display: inline-flex;
    vertical-align: middle;
    align-items: center;
    height: 100%;
    flex: 1;
    user-select: none;
    border: 1px solid var(--fay-border);
    border-radius: var(--fay-radius);
    transition: border 0.2s ease 0s, color 0.2s ease 0s;
    padding: 0;
    &.hover {
      border-color: var(--accents-5);
    }
    &.disabled {
      cursor: not-allowed;
      background-color: var(--accents-1);
      border-color: var(--accents-2);
    }
  }

  .input_wrapper {
    input {
      margin: 4px 10px;
      box-shadow: none;
      background-color: transparent;
      border: none;
      outline: none;
      border-radius: 0;
      width: 100%;
      min-width: 0;
      height: 100%;
      -webkit-appearance: none;
      color: var(--primary-foreground);
      &::placeholder {
        color: var(--accents-3);
      }
      &.disabled {
        cursor: not-allowed;
      }
    }
  }
  label {
    display: block;
    font-weight: normal;
    padding: 0 0 0 1px;
    font-size: 16px;
    line-height: 1.5;
    color: var(--accents-6);
    margin-bottom: var(--fay-gap-half);
    &:first-child {
      margin-top: 0;
    }
    &:last-child {
      margin-bottom: 0;
    }
  }
}
