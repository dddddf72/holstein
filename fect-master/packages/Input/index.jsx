import { computed, ref, watchEffect } from 'vue'
import { createNameSpace, theme, validator } from '../utils'
import ClearableIcon from './clearable-icon'
import PasswordIcon from './password-icon'
import IconContent from './icon-content'
import InputLabel from './input-label'
import InputBlockLabel from './input-block-label'
import './input.less'

const [createComponent] = createNameSpace('Input')

const { normalSizes } = theme

const queryInputSize = (size) => {
  const sizesPool = {
    mini: {
      heightRatio: '1.313',
      fontSize: '12px',
    },
    small: {
      heightRatio: '1.5',
      fontSize: '12px',
    },
    medium: {
      heightRatio: '1.687',
      fontSize: '14px',
    },
    large: {
      heightRatio: '1.875',
      fontSize: '16px',
    },
  }

  return sizesPool[size]
}

export default createComponent({
  props: {
    modelValue: {
      type: [String, Number],
      default: '',
    },
    type: { type: String, default: 'text' },
    placeholder: String,
    size: {
      type: String,
      default: 'medium',
      validator: validator.enums(normalSizes),
    },
    autocomplete: String,
    readonly: Boolean,
    disabled: Boolean,
    clearable: Boolean,
    prefix: [String, Number],
    suffix: [String, Number],
  },
  emits: ['change', 'blur', 'focus', 'clearClick', 'update:modelValue'],
  setup(props, { attrs, slots, emit }) {
    const inputRef = ref(null)
    const hover = ref(false)
    const selfType = ref(props.type)
    const passwordVisible = ref(false)
    const fontSize = ref(null)
    const heightRatio = ref(null)
    const hasPrefix = ref(!!props.prefix)
    const hasSuffix = ref(!!props.suffix)

    const setHoverable = (pre) => (hover.value = pre)

    watchEffect(() => {
      const { heightRatio: Ratio, fontSize: Size } = queryInputSize(props.size)
      heightRatio.value = Ratio
      fontSize.value = Size
    })

    const updatelValue = (val) => {
      if (props.type === 'number') {
        val = Number(val)
      }
      if (inputRef.value && val !== inputRef.value.value) {
        inputRef.value.value = val
      }
      if (val !== props.modelValue) {
        emit('update:modelValue', val)
      }
    }

    const InputHandler = (e) => updatelValue(e.target.value)

    const focusHandler = (e) => {
      setHoverable(true)
      emit('focus', e)
    }

    const blurHandler = (e) => {
      setHoverable(false)
      emit('blur', e)
    }

    const changeHandler = (e) => {
      if (props.disabled || props.readonly) return
      emit('change', e)
    }

    const clearHandler = (e) => {
      updatelValue('')
      emit('clearClick', e)
    }

    /**
     *control clearable icon visible
     */
    const setClearable = computed(() => Boolean(props.modelValue !== ''))

    const renderInput = () => {
      const InputProps = {
        ref: inputRef,
        value: props.modelValue,
        disabled: props.disabled,
        readOnly: props.readonly,
        placeholder: props.placeholder,
        autocomplete: props.autocomplete,
        onInput: InputHandler,
        onFocus: focusHandler,
        onBlur: blurHandler,
        onChange: changeHandler,
        ...attrs,
      }

      const setPasswordVisible = () => {
        passwordVisible.value = !passwordVisible.value
        if (passwordVisible.value) return (selfType.value = 'text')
        return (selfType.value = 'password')
      }

      return (
        <>
          <input
            type={selfType.value}
            class={`${props.disabled ? 'disabled' : ''}`}
            style={{ fontSize: fontSize.value }}
            {...InputProps}
          />
          {props.clearable ? (
            <ClearableIcon
              visible={setClearable.value}
              disabled={props.disabled || props.readonly}
              onClick={clearHandler}
            />
          ) : (
            ''
          )}
          {props.type === 'password' ? (
            <IconContent
              onClick={setPasswordVisible}
              clickable={props.disabled || props.readonly}
            >
              <PasswordIcon visible={passwordVisible.value} />
            </IconContent>
          ) : (
            ''
          )}
        </>
      )
    }

    const haslabel = computed(() => {
      const prefixStyle = {
        borderTopLeftRadius: 0,
        borderBottomLeftRadius: 0,
      }
      const suffixStyle = {
        borderTopRightRadius: 0,
        borderBottomRightRadius: 0,
      }

      if (hasPrefix.value && hasSuffix.value) {
        return Object.assign({}, prefixStyle, suffixStyle)
      }
      if (hasPrefix.value) {
        return prefixStyle
      }
      if (hasSuffix.value) {
        return suffixStyle
      }
      return ''
    })

    return () => (
      <div class="fect-input" style={{ '--heightRatio': heightRatio.value }}>
        {slots.default && <InputBlockLabel v-slots={slots} />}
        <div class={'input_container'}>
          {hasPrefix.value && (
            <InputLabel fontSize={fontSize.value}>{props.prefix}</InputLabel>
          )}
          <div
            class={`input_wrapper ${hover.value ? 'hover' : ''} ${
              props.disabled ? 'disabled' : ''
            }`}
            style={haslabel.value}
          >
            {renderInput()}
          </div>
          {hasSuffix.value && (
            <InputLabel fontSize={fontSize.value} isRight>
              {props.suffix}
            </InputLabel>
          )}
        </div>
      </div>
    )
  },
})
