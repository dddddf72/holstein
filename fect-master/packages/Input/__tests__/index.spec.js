import { mount } from '@vue/test-utils'
import { ref } from 'vue'
import Input from '../index'

describe('Input', () => {
  it('should be render as a element', () => {
    const wrapper = mount(<Input />)
    expect(() => wrapper.unmount()).not.toThrow()
  })
  it('should be support different sizes', () => {
    const wrapper = mount({
      setup() {
        return {
          sizes: ['mini', 'small', 'medium', 'large'],
        }
      },
      render() {
        return (
          <>
            {this.sizes.map((size, idx) => (
              <Input key={idx} size={size} />
            ))}
          </>
        )
      },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should work with label', () => {
    const wrapper = mount(
      <Input>
        <span>Test</span>
      </Input>,
    )
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('label').html()).toContain(
      '<label><span>Test</span></label>',
    )
  })
  it('should support prefix label', () => {
    const wrapper = mount(<Input prefix="Test"></Input>)
    expect(wrapper.find('.input-label').text()).toBe('Test')
  })

  it('should support suffix label', () => {
    const wrapper = mount(<Input suffix="Test"></Input>)
    expect(wrapper.find('.input-label').classes('suffix')).toBe(true)
  })

  it('should set input from value', () => {
    const wrapper = mount(Input, {
      props: {
        type: 'text',
        modelValue: '',
      },
    })
    const input = wrapper.find('input')
    input.element.value = '1'
    input.trigger('input')
    expect(wrapper.emitted('update:modelValue')[0][0]).toEqual('1')
  })

  it('should support clearable', () => {
    const wrapper = mount(<Input clearable />)
    expect(wrapper.find('.input__clear-icon')).toBeTruthy()
  })

  it('should support password Icon', () => {
    const wrapper = mount(<Input type="password" />)
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('when prpos type is number emit value should be number type', () => {
    const wrapper = mount(Input, {
      props: {
        modelValue: 0,
        type: 'number',
      },
    })
    const input = wrapper.find('input')
    input.element.value = 2
    input.trigger('input')
    expect(wrapper.emitted('update:modelValue')[0][0]).toEqual(2)
  })
  
})
