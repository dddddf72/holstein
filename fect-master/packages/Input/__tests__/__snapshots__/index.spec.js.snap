// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should be support different sizes 1`] = `
<div class="fect-input" style="--heightRatio: 1.313;">
  <!---->
  <div class="input_container">
    <!---->
    <div class="input_wrapper  "><input type="text" class="" style="font-size: 12px;"></div>
    <!---->
  </div>
</div>
<div class="fect-input" style="--heightRatio: 1.5;">
  <!---->
  <div class="input_container">
    <!---->
    <div class="input_wrapper  "><input type="text" class="" style="font-size: 12px;"></div>
    <!---->
  </div>
</div>
<div class="fect-input" style="--heightRatio: 1.687;">
  <!---->
  <div class="input_container">
    <!---->
    <div class="input_wrapper  "><input type="text" class="" style="font-size: 14px;"></div>
    <!---->
  </div>
</div>
<div class="fect-input" style="--heightRatio: 1.875;">
  <!---->
  <div class="input_container">
    <!---->
    <div class="input_wrapper  "><input type="text" class="" style="font-size: 16px;"></div>
    <!---->
  </div>
</div>
`;

exports[`Input should support password Icon 1`] = `
<div class="fect-input" style="--heightRatio: 1.687;">
  <!---->
  <div class="input_container">
    <!---->
    <div class="input_wrapper  "><input type="password" class="" style="font-size: 14px;">
      <div class="input-icon" style="cursor: pointer;"><svg viewBox="0 0 24 24" width="16" height="16" stroke="currentColor" strokeWidth="1.5" strokeLinecap="round" strokeLinejoin="round" fill="none" shapeRendering="geometricPrecision" style="color: currentColor;">
          <path d="M1 12s4-8 11-8 11 8 11 8-4 8-11 8-11-8-11-8z"></path>
          <circle cx="12" cy="12" r="3"></circle>
        </svg></div>
    </div>
    <!---->
  </div>
</div>
`;

exports[`Input should work with label 1`] = `
<div class="fect-input" style="--heightRatio: 1.687;"><label><span>Test</span></label>
  <div class="input_container">
    <!---->
    <div class="input_wrapper  "><input type="text" class="" style="font-size: 14px;"></div>
    <!---->
  </div>
</div>
`;
